[iOS Mail] Selection does not change when tapping in compose when keyboard is not already shown

whsieh opened the pull request:
#### 8d38f657f8eeec2823a753ac4b509cae9f4b70f1
<pre>
[iOS Mail] Selection does not change when tapping in compose when keyboard is not already shown
<a href="https://bugs.webkit.org/show_bug.cgi?id=243356">https://bugs.webkit.org/show_bug.cgi?id=243356</a>
rdar://96609950

Reviewed by Aditya Keerthi.

When tapping in Mail compose when the keyboard is not already shown, we first
`-becomeFirstResponder` underneath `-_singleTapRecognized:` before proceeding to dispatch synthetic
click events to the page. Note that:

1.  The entire web view is treated as editable via `WKWebView` SPI, and
2.  Mail implements the `-_webView:decidePolicyForFocusedElement:` input delegate method and returns
    `_WKFocusStartsInputSessionPolicyAllow`

...which causes us to show the keyboard earlier than usual (before actually completing the synthetic
click and dispatching mouse events), due to setting the focused element underneath activity state
flag changes (specifically, `IsFocused`). This, in turn, causes us to send IPC from the UI process
back to the web process indicating that the keyboard is shown, which causes us to *not* set the
selection underneath WebKit2 editor client code, due to the fact that
`isShowingInputViewForFocusedElement()` is true:

```
bool WebEditorClient::shouldAllowSingleClickToChangeSelection(…) const
{
    …
    return !editableRoot || editableRoot != targetNode.rootEditableElement() || !m_page-&gt;isShowingInputViewForFocusedElement();
}
```

To fix this (and allow the selection to change when tapping), we add some new state flags to
`WebPage` to only consider whether or not the keyboard *was* shown during the potential tap when
determining whether a synthetic click should change the selection.

Test: editing/selection/ios/tap-to-set-selection-in-editable-web-view.html

* LayoutTests/editing/selection/ios/tap-to-set-selection-in-editable-web-view-expected.txt: Added.
* LayoutTests/editing/selection/ios/tap-to-set-selection-in-editable-web-view.html: Added.

Add a layout test to exercise the bug; see above for more details.

* Source/WebKit/WebProcess/WebCoreSupport/ios/WebEditorClientIOS.mm:
(WebKit::WebEditorClient::shouldAllowSingleClickToChangeSelection const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::potentialTapAtPosition):
(WebKit::WebPage::shouldAllowSingleClickToChangeSelection):

Move this logic out of `WebEditorClient` and into `WebPage` as a helper method; in the case where
we&apos;re completing a synthetic click (i.e. dispatching mousedown and click events when tapping), we
consult `m_wasShowingInputViewForFocusedElementDuringLastPotentialTap` rather than the current state
of `m_isShowingInputViewForFocusedElement`.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::focusStartsInputSessionPolicy const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView _webView:decidePolicyForFocusedElement:]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::focusStartsInputSessionPolicy):
(WTR::TestController::platformResetStateToConsistentValues):

Implement the input delegate method `-_webView:decidePolicyForFocusedElement:`, and add a new layout
test option to influence the return type of this delegate method.

Canonical link: <a href="https://commits.webkit.org/252972@main">https://commits.webkit.org/252972@main</a>
</pre>

pxlcoder commented on Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
```diff
@@ -119,6 +118,7 @@ - (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguratio
         [center addObserver:self selector:@selector(_didDismissPopover) name:@"UIPopoverControllerDidDismissPopoverNotification" object:nil];
         self.UIDelegate = self;
         self._inputDelegate = self;
+        self.focusStartsInputSessionPolicy = _WKFocusStartsInputSessionPolicyAuto;
```
Do we need to reset this in `TestController::platformResetStateToConsistentValues`?

whsieh replied:
Hm...I think the change in https://github.com/WebKit/WebKit/pull/2864/files#diff-0a820bf40aebcff03905d48e022f8c7c31443aeec49d04bbcefb5d2e32818488R273 should cover this?

pxlcoder replied:
Was tripped up by the UI, which was showing this underneath `restorePortraitOrientationIfNeeded`. Looks good!

pxlcoder reviewed (approved):
I think the PR title needs to be changed to remove the `is`, or change the `with` to `when`.

whsieh committed:
[iOS Mail] Selection does not change when tapping in compose when keyboard is not already shown
https://bugs.webkit.org/show_bug.cgi?id=243356
rdar://96609950

Reviewed by Aditya Keerthi.

When tapping in Mail compose when the keyboard is not already shown, we first
`-becomeFirstResponder` underneath `-_singleTapRecognized:` before proceeding to dispatch synthetic
click events to the page. Note that:

1.  The entire web view is treated as editable via `WKWebView` SPI, and
2.  Mail implements the `-_webView:decidePolicyForFocusedElement:` input delegate method and returns
    `_WKFocusStartsInputSessionPolicyAllow`

...which causes us to show the keyboard earlier than usual (before actually completing the synthetic
click and dispatching mouse events), due to setting the focused element underneath activity state
flag changes (specifically, `IsFocused`). This, in turn, causes us to send IPC from the UI process
back to the web process indicating that the keyboard is shown, which causes us to *not* set the
selection underneath WebKit2 editor client code, due to the fact that
`isShowingInputViewForFocusedElement()` is true:

```
bool WebEditorClient::shouldAllowSingleClickToChangeSelection(…) const
{
    …
    return !editableRoot || editableRoot != targetNode.rootEditableElement() || !m_page->isShowingInputViewForFocusedElement();
}
```

To fix this (and allow the selection to change when tapping), we add some new state flags to
`WebPage` to only consider whether or not the keyboard *was* shown during the potential tap when
determining whether a synthetic click should change the selection.

Test: editing/selection/ios/tap-to-set-selection-in-editable-web-view.html

* LayoutTests/editing/selection/ios/tap-to-set-selection-in-editable-web-view-expected.txt: Added.
* LayoutTests/editing/selection/ios/tap-to-set-selection-in-editable-web-view.html: Added.

Add a layout test to exercise the bug; see above for more details.

* Source/WebKit/WebProcess/WebCoreSupport/ios/WebEditorClientIOS.mm:
(WebKit::WebEditorClient::shouldAllowSingleClickToChangeSelection const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::potentialTapAtPosition):
(WebKit::WebPage::shouldAllowSingleClickToChangeSelection):

Move this logic out of `WebEditorClient` and into `WebPage` as a helper method; in the case where
we're completing a synthetic click (i.e. dispatching mousedown and click events when tapping), we
consult `m_wasShowingInputViewForFocusedElementDuringLastPotentialTap` rather than the current state
of `m_isShowingInputViewForFocusedElement`.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::focusStartsInputSessionPolicy const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView _webView:decidePolicyForFocusedElement:]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::focusStartsInputSessionPolicy):
(WTR::TestController::platformResetStateToConsistentValues):

Implement the input delegate method `-_webView:decidePolicyForFocusedElement:`, and add a new layout
test option to influence the return type of this delegate method.

Canonical link: https://commits.webkit.org/252972@main
